File: Stage/C++/Ui/legende.cpp
#include "legende.h"
#include "ui_legende.h"

// classe qui herite de vue ca fonction refresh sera appeler a chaque fois que les donnees d'un model (groupes, stations,parametres )changent

/*--------------------------------CONSTRUCTEURS--------------------------------*/
Legende::Legende(QWidget *parent) :
    QWidget(parent),
    ui(new Ui::Legende)
{
    ui->setupUi(this);
}

Legende::Legende(Stations *s, Parametres *p, Groupes *g, QWidget *parent):
    QWidget(parent),
    ui(new Ui::Legende),
    modelStation(s),
    modelParametre(p),
    modelGroupe(g)
{
    modelStation->abonne(this);
    modelParametre->abonne(this);

    ui->setupUi(this);

    refresh();
}

/*--------------------------------DESTRUCTEURS--------------------------------*/
Legende::~Legende()
{
    delete ui;
}

/*--------------------------------FONCTIONS PUBLICS--------------------------------*/

// fonction qui permet de rafraichir les differents elements de la fenetre
void Legende::refresh(){
    init_table();
    QStandardItemModel *modelTabParametre = qobject_cast<QStandardItemModel*>(ui->tableViewParametre->model());
    QStandardItemModel *modelTabStation = qobject_cast<QStandardItemModel*>(ui->tableViewStation->model());

    // on connect le signal dataChanged des listeView a la fonction tableViewStation_dataChanged
    connect(modelTabStation, SIGNAL(dataChanged(QModelIndex,QModelIndex)),this,SLOT(tableViewStation_dataChanged(QModelIndex)));
    connect(modelTabParametre, SIGNAL(dataChanged(QModelIndex,QModelIndex)),this,SLOT(tableViewParametre_dataChanged(QModelIndex)));
}

/*--------------------------------FONCTIONS PRIVATE SLOTS--------------------------------*/

// fonction appele quand on double clique sur la tableViewStation (ouvre color dialog ou enregistre la valeur de la cellule)
void Legende::on_tableViewStation_clicked(const QModelIndex &index)
{
    ui->tableViewStation->clearSelection();
    // si l'item cliquer est sur la deuxieme colonne c'est une couleur on ouvre donc la color dialog sinon on enregistre la valeur de la cellule
    if(index.column()==1){
        QStandardItemModel *modelTabStation = qobject_cast<QStandardItemModel*>(ui->tableViewStation->model());
        QStandardItem *item= modelTabStation->item(index.row(),1);
        colorChange(item);

        QString nomStation = modelTabStation->item(index.row(),0)->text();
        QColor newColor = item->background().color();
        QString couleur = QString::number(newColor.red())+","+QString::number(newColor.green())+","+QString::number(newColor.blue());
        modelStation->setCouleurStation(nomStation,couleur);
    }else if(index.column()==3){
        QStandardItemModel *modelTabStation = qobject_cast<QStandardItemModel*>(ui->tableViewStation->model());
        QString nomStation = modelTabStation->item(index.row(),0)->text();

        if(modelGroupe->getListNomParent().contains(nomStation)){
            Groupe *groupe = modelGroupe->getGroupe(nomStation);
            modelGroupe->removeGroupe(groupe);
        }
        modelGroupe->removeElementFils(modelStation->getStation(nomStation));

        modelStation->removeStation(nomStation);
    }else
        currentItemData=index.data().toString();
}

// fonction appele quand on double clique sur la tableViewParametre (ouvre color dialog ou enregistre la valeur de la cellule)
void Legende::on_tableViewParametre_clicked(const QModelIndex &index)
{
    ui->tableViewParametre->clearSelection();
    // si l'item cliquer est sur la deuxieme colonne c'est une couleur on ouvre donc la color dialog sinon on enregistre la valeur de la cellule
    if(index.column()==1){
        QStandardItemModel *modelTabParametre = qobject_cast<QStandardItemModel*>(ui->tableViewParametre->model());
        QStandardItem *item= modelTabParametre->item(index.row(),1);
        colorChange(item);

        QString nomParametre = modelTabParametre->item(index.row(),0)->text();
        QColor newColor = item->background().color();
        QString couleur = QString::number(newColor.red())+","+QString::number(newColor.green())+","+QString::number(newColor.blue());
        modelParametre->setCouleurParametre(nomParametre,couleur);

    }else if(index.column()==3){
        QStandardItemModel *modelTabParametre = qobject_cast<QStandardItemModel*>(ui->tableViewParametre->model());
        QString nomParametre = modelTabParametre->item(index.row(),0)->text();

        if(modelGroupe->getListNomParent().contains(nomParametre)){
            Groupe *groupe = modelGroupe->getGroupe(nomParametre);
            modelGroupe->removeGroupe(groupe);
        }
        modelGroupe->removeElementFils(modelParametre->getParametre(nomParametre));
        modelParametre->removeParametre(nomParametre);
    }else
        currentItemData=index.data().toString();

}

//fonction appele quand les donnees de la tableView changent (change le nom du parametre, le nom de la categorie)
void Legende::tableViewStation_dataChanged(const QModelIndex & topLeft){
    if(topLeft.column()==0){
        modelStation->setNomStation(currentItemData,topLeft.data().toString());
    }
    if(topLeft.column()==2){
        QStandardItemModel *modelTabStation = qobject_cast<QStandardItemModel*>(ui->tableViewStation->model());
        QModelIndex indexStation = modelTabStation->index(topLeft.row(),0);
        modelStation->setCategorie(indexStation.data().toString(),topLeft.data().toString());
    }
}

//fonction appele quand les donnees de la tableView changent (change le nom du parametre, le nom de la categorie)
void Legende::tableViewParametre_dataChanged(const QModelIndex & topLeft){
    if(topLeft.column()==0){
        modelParametre->setNomParametre(currentItemData,topLeft.data().toString());
    }
    if(topLeft.column()==2){
        QStandardItemModel *modelTabParametre = qobject_cast<QStandardItemModel*>(ui->tableViewParametre->model());
        QModelIndex indexParametre = modelTabParametre->index(topLeft.row(),0);
        modelParametre->setCategorie(indexParametre.data().toString(),topLeft.data().toString());
    }
}

/*--------------------------------FONCTIONS PRIVATE--------------------------------*/

// fonction qui permet d'initialiser les tableaux
void Legende::init_table(){

    ui->tableViewStation->verticalHeader()->setVisible(false);
    ui->tableViewStation->setSelectionMode(QAbstractItemView::SingleSelection);

    ui->tableViewParametre->verticalHeader()->setVisible(false);
    ui->tableViewParametre->setSelectionMode(QAbstractItemView::SingleSelection);

    QStringList header;

    header.append("Nom");
    header.append("Code Couleur");
    header.append("Catégorie");
    header.append("Supprimer");

    init_TabParametre(header);
    init_TabStation(header);
}

// fonction qui permet d'initialiser le tableau parametre
void Legende::init_TabParametre(QStringList header){
    QList<Element*> parametres = modelParametre->getListParametre();
    QStandardItemModel *modelTabParametre = new QStandardItemModel();
    ui->tableViewParametre->setModel(modelTabParametre);

    modelTabParametre->setHorizontalHeaderLabels(header);

    foreach(Element* parametre,parametres){
        QStandardItem *itemNom = new QStandardItem(parametre->getNom());
        QString codeCouleur = parametre->getCouleur();

        QStringList rgb = codeCouleur.split(",");
        QColor couleur(rgb.at(0).toInt(),rgb.at(1).toInt(),rgb.at(2).toInt(),255);
        QBrush brush(couleur);

        QStandardItem *itemCouleur = new QStandardItem();
        itemCouleur->setBackground(brush);
        itemCouleur->setEditable(false);

        QStandardItem *itemCategorie = new QStandardItem(parametre->getCategorie());

        QStandardItem *itemSupprimer = new QStandardItem("X");
        itemSupprimer->setTextAlignment(Qt::AlignHCenter);
        itemSupprimer->setEditable(false);
        itemSupprimer->setSelectable(false);

        QList<QStandardItem*> items;
        items.append(itemNom);
        items.append(itemCouleur);
        items.append(itemCategorie);
        items.append(itemSupprimer);

        modelTabParametre->appendRow(items);
    }

}

// fonction qui permet d'initialiser le tableau station
void Legende::init_TabStation(QStringList header){
    QList<Element*> stations = modelStation->getListStation();
    QStandardItemModel *modelTabStation = new QStandardItemModel();
    ui->tableViewStation->setModel(modelTabStation);


    modelTabStation->setHorizontalHeaderLabels(header);

    foreach(Element* station,stations){
        QStandardItem *itemNom = new QStandardItem(station->getNom());
        QString codeCouleur = station->getCouleur();

        QStringList rgb = codeCouleur.split(",");
        QColor couleur(rgb.at(0).toInt(),rgb.at(1).toInt(),rgb.at(2).toInt(),255);
        QBrush brush(couleur);

        QStandardItem *itemCouleur = new QStandardItem();
        itemCouleur->setBackground(brush);
        itemCouleur->setEditable(false);

        QStandardItem *itemCategorie = new QStandardItem(station->getCategorie());

        QStandardItem *itemSupprimer = new QStandardItem("X");
        itemSupprimer->setTextAlignment(Qt::AlignHCenter);
        itemSupprimer->setEditable(false);
        itemSupprimer->setSelectable(false);

        QList<QStandardItem*> items;
        items.append(itemNom);
        items.append(itemCouleur);
        items.append(itemCategorie);
        items.append(itemSupprimer);

        modelTabStation->appendRow(items);
    }


}

// fonction qui permet de changer la couleur d'un element
void Legende::colorChange(QStandardItem *item){
    QColorDialog colorDialog;
    QBrush brush = item->background();

    QColor couleur = colorDialog.getColor(brush.color());

    if(couleur.isValid())
        brush.setColor(couleur);

    item->setBackground(brush);
}

File: Stage/Python/Fenetres/Legende.py
from PyQt5 import QtCore, QtGui, QtWidgets
from UI.UIlegende import Ui_Form

class F_Legende(QtWidgets.QWidget, Ui_Form):
 
    def __init__(self, xmlFile, model, parent=None):
        super(F_Legende, self).__init__(parent)
        self.model = model
        
        ## on abonne la fenetre au model pour quel recoive les modifications
        self.model.get("stations").abonne(self)
        self.model.get("parametres").abonne(self)
        
        self.setupUi(self)
        self.refresh()
    
    def refresh(self):
        self.init_Table()
        
        ## on connect l'evenement clicked des deux tableaux a la fonction getSelection
        self.tableViewStation.clicked.connect(self.getSelectionStation)
        self.tableViewParametre.clicked.connect(self.getSelectionParametre)
        
        ## on connect l'evenement doubleclicked des deux tableaux a la fonction getDoubleClicked
        self.tableViewStation.doubleClicked.connect(self.getDoubleClicked)
        self.tableViewParametre.doubleClicked.connect(self.getDoubleClicked)

        ## on connect l'evenement dataChanged des deux tableaux a la fonction getDataChanged
        self.tableViewStation.model().dataChanged.connect(self.getDataChangedStation)        
        self.tableViewParametre.model().dataChanged.connect(self.getDataChangedParametre)

        
    def init_Table(self):
        ## on recupere les models
        modelSta = self.model.get("stations")
        modelPar = self.model.get("parametres")
        
        ## on recupere la liste des stations       
        stations = modelSta.getListStation()

        ## on recupere la liste des parametres
        parametres = modelPar.getListParametre()
        
        listeHeader = ["Nom","Code Couleur"]

        ## on place les headers des deux model de tableau
        self.modelTabStation = QtGui.QStandardItemModel()
        self.modelTabStation.setHorizontalHeaderLabels(listeHeader)

        self.modelTabParametre = QtGui.QStandardItemModel()
        self.modelTabParametre.setHorizontalHeaderLabels(listeHeader)

        ## on lie les model aux vus
        self.tableViewStation.setModel(self.modelTabStation)
        self.tableViewParametre.setModel(self.modelTabParametre)

        ## on enleve les header verticale et un met les tableau en selection unique
        self.tableViewParametre.verticalHeader().setVisible(0)
        self.tableViewParametre.setSelectionMode(QtWidgets.QAbstractItemView.SingleSelection)

        self.tableViewStation.verticalHeader().setVisible(0)
        self.tableViewStation.setSelectionMode(QtWidgets.QAbstractItemView.SingleSelection)  

        ## on initialise le contenu des tableaux
        self.init_TabStations(stations, modelSta)
        self.init_TabParametres(parametres, modelPar) 

## fonction qui initialise le contenu du tableau station       
    def init_TabStations(self, stations , modelSta):
        
        for i in range(0,len(stations)):
           station = stations[i]
           itemNom = QtGui.QStandardItem(station.getNom())

           ## on recupere la couleur sous la forme 255,255,255
           couleurStation = station.getCouleur()

           ## on split la chaine pour avoir les valeurs separement
           codeRgb = couleurStation.split(",")

           ##on creer la couleur et le brush pour l'item
           couleur = QtGui.QColor(int(codeRgb[0]),int(codeRgb[1]),int(codeRgb[2]),255)
           brush = QtGui.QBrush(couleur)
           
           itemCouleur = QtGui.QStandardItem()
           itemCouleur.setBackground(brush)
           itemCouleur.setEditable(0)

           self.modelTabStation.insertRow(i,[itemNom,itemCouleur])

## fonction qui initialise le contenu du tableau parametre  
    def init_TabParametres(self, parametres, modelPar):
        for i in range(0,len(parametres)):
           parametre = parametres[i] 
           itemNom = QtGui.QStandardItem(parametre.getNom())

           ## on recupere la couleur sous la forme 255,255,255
           couleurParametre = parametre.getCouleur()

           ## on split la chaine pour avoir les valeurs separement
           codeRgb = couleurParametre.split(",")

           ## on creer la couleur et le brush pour l'item
           couleur = QtGui.QColor(int(codeRgb[0]),int(codeRgb[1]),int(codeRgb[2]),255)
           brush = QtGui.QBrush(couleur)
           
           itemCouleur = QtGui.QStandardItem()
           itemCouleur.setBackground(brush)
           itemCouleur.setEditable(0)

           ## on ajoute les item dans le model du tableau
           self.modelTabParametre.insertRow(i,[itemNom,itemCouleur])

## fonction appele quand l'on double clique dans le tableau
## on change la valeur de cureentItemData qui contient les donnees de l'objet sur lequel on a double clicquer
    def getDoubleClicked(self, item):
        if(item.column()==0):
            self.currentItemData = item.data()
            
## fonction appele quand une donne change dans le tableau station
    def getDataChangedStation(self,item):
        model = self.model.get("stations")
        
        if(item.column()==0):
           model.setNomStation(self.currentItemData,item.data())
           
## fonction appeler quand une donne change dans le tableau parametre           
    def getDataChangedParametre(self,item):
        model = self.model.get("parametres")
        
        if(item.column()==0):
           model.setNomParametre(self.currentItemData,item.data())
           
## fonction appele quand l'utilisateur clique sur un case de l'onglet Station
    def getSelectionStation(self,item):

        ## on regarde si l'item selectionne est dans la colonnes couleur
        if(item.column()==1):
            ## on recupere le model de l'item pour le changer
            itemMod = self.modelTabStation.item(item.row(),1)

            ## on change la couleur de l'item
            self.colorChange(itemMod)

        ## on deselectionne tous les items situees sur ou dans le rectangle(le tableau)  
        self.tableViewStation.setSelection(QtCore.QRect(0,0,2,item.row()+1),QtCore.QItemSelectionModel.Clear)

## fonction appele quand l'utilisateur clique sur un case de l'onglet Parametre    
    def getSelectionParametre(self, item):
        ## meme fonctionnement que la fonction getSelectionStation
        if(item.column()==1):
            itemMod = self.modelTabParametre.item(item.row(),1)
            self.colorChange(itemMod)
            
        self.tableViewParametre.setSelection(QtCore.QRect(0,0,2,item.row()+1),QtCore.QItemSelectionModel.Clear)

## fonction qui change la couleur de fond de l'item selectioner
## utilise un colordialog pour la selection de la couleur
    def colorChange(self, itemMod):
        colorDialog = QtWidgets.QColorDialog(self)
        brush = itemMod.background()

        color = colorDialog.getColor(brush.color())
        brush.setColor(color)
        itemMod.setBackground(brush)
                     
